File: apps/web/src/services/api/events.ts
/**
 * Events API client (GET /api/v1/events, GET /api/v1/events/{id}).
 */

const BASE = import.meta.env.VITE_API_BASE_URL ?? '/api/v1'

export interface EventListItem {
  id: number
  event_type: string
  country_id: number | null
  actor: string | null
  impact_type: string | null
  occurred_at: string
  source_summary: string | null
  confidence: number | null
  extracted_at: string
}

export interface EventListResponse {
  items: EventListItem[]
  total: number
}

export interface NewsSource {
  id: number
  source: string
  url: string | null
  title: string | null
}

export interface EventDetail {
  id: number
  event_type: string
  country_id: number | null
  actor: string | null
  impact_type: string | null
  occurred_at: string
  source_summary: string | null
  confidence: number | null
  extracted_at: string
  source: NewsSource | null
  metadata_?: Record<string, unknown> | null
}

export interface EventListParams {
  type?: string
  country?: number
  from_date?: string
  to_date?: string
  limit?: number
  offset?: number
}

export async function getEvents (params?: EventListParams): Promise<EventListResponse> {
  const search = new URLSearchParams()
  if (params?.type != null) search.set('type', params.type)
  if (params?.country != null) search.set('country', String(params.country))
  if (params?.from_date != null) search.set('from_date', params.from_date)
  if (params?.to_date != null) search.set('to_date', params.to_date)
  if (params?.limit != null) search.set('limit', String(params.limit))
  if (params?.offset != null) search.set('offset', String(params.offset))
  const qs = search.toString()
  const url = `${BASE}/events${qs ? `?${qs}` : ''}`
  const res = await fetch(url)
  if (!res.ok) throw new Error(await res.text() || `HTTP ${res.status}`)
  return res.json()
}

export async function getEvent (id: number): Promise<EventDetail> {
  const res = await fetch(`${BASE}/events/${id}`)
  if (!res.ok) throw new Error(await res.text() || `HTTP ${res.status}`)
  return res.json()
}

File: apps/api/src/api/events.py
"""Event API: GET /events, GET /events/{id} (contracts/api.md)."""

from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from core.database import get_async_session
from models.event import Event
from services.event_service import EventFilters, EventService


router = APIRouter()


class EventListItem(BaseModel):
    id: int
    event_type: str
    country_id: int | None
    actor: str | None
    impact_type: str | None
    occurred_at: datetime
    source_summary: str | None
    confidence: float | None
    extracted_at: datetime

    model_config = {"from_attributes": True}


class EventListResponse(BaseModel):
    items: list[EventListItem]
    total: int


class NewsSource(BaseModel):
    id: int
    source: str
    url: str | None
    title: str | None

    model_config = {"from_attributes": True}


class EventDetailResponse(BaseModel):
    id: int
    event_type: str
    country_id: int | None
    actor: str | None
    impact_type: str | None
    occurred_at: datetime
    source_summary: str | None
    confidence: float | None
    extracted_at: datetime
    source: NewsSource | None = None
    metadata_: dict | None = None

    model_config = {"from_attributes": True}


def _event_to_detail(event: Event) -> dict:
    data: dict = {
        "id": event.id,
        "event_type": event.event_type,
        "country_id": event.country_id,
        "actor": event.actor,
        "impact_type": event.impact_type,
        "occurred_at": event.occurred_at,
        "source_summary": event.source_summary,
        "confidence": event.confidence,
        "extracted_at": event.extracted_at,
        "metadata_": event.metadata_,
    }
    if event.news is not None:
        data["source"] = NewsSource(
            id=event.news.id,
            source=event.news.source,
            url=event.news.url,
            title=event.news.title,
        )
    else:
        data["source"] = None
    return data


@router.get("", response_model=EventListResponse)
async def list_events(
    type: str | None = Query(None, alias="type", description="event_type filter"),
    country: int | None = Query(None, description="country_id filter"),
    from_date: datetime | None = Query(None, description="occurred_at from"),
    to_date: datetime | None = Query(None, description="occurred_at to"),
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
    session: AsyncSession = Depends(get_async_session),
):
    """List events with optional filters (type, country, from_date, to_date, limit, offset)."""
    filters = EventFilters(
        type=type,
        country=country,
        from_date=from_date,
        to_date=to_date,
        limit=limit,
        offset=offset,
    )
    svc = EventService(session)
    items, total = await svc.list(filters)
    return EventListResponse(
        items=[EventListItem.model_validate(e) for e in items],
        total=total,
    )


@router.get("/{event_id}", response_model=EventDetailResponse)
async def get_event(
    event_id: int,
    session: AsyncSession = Depends(get_async_session),
):
    """Get single event by id; includes source (news) and extracted attributes."""
    svc = EventService(session)
    event = await svc.get_by_id_with_source(event_id)
    if event is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Event not found",
        )
    return EventDetailResponse(**_event_to_detail(event))